src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
+            size += OBJ_ID_SIZE + BYTE_SIZE;
+            // ik == null for instance fields
+            if (ik != null) {
+                // static field
+                size += calculateFieldDumpRecordSize(field, ik.getJavaMirror());
+            }
         }
+        return size;
+    }
+
+    private int calculateFieldDumpRecordSize(Field field, Oop oop) throws IOException {
+        char typeCode = (char) field.getSignature().getByteAt(0);
+        int size = 0;
+        switch (typeCode) {
+            case JVM_SIGNATURE_BOOLEAN:
+            case JVM_SIGNATURE_BYTE:
+                size++;
+                break;
+            case JVM_SIGNATURE_CHAR:
+            case JVM_SIGNATURE_SHORT:
+                size += 2;
+                break;
+            case JVM_SIGNATURE_INT:
+            case JVM_SIGNATURE_FLOAT:
+                size += 4;
+                break;
+            case JVM_SIGNATURE_CLASS:
+            case JVM_SIGNATURE_ARRAY:
+                size += OBJ_ID_SIZE;
+                break;
+            case JVM_SIGNATURE_LONG:
+            case JVM_SIGNATURE_DOUBLE:
+                size += 8;
+                break;
+            default:
+                throw new RuntimeException("should not reach here");
+        }
+        return size;
+    }
+
+    private int calculateClassInstanceDumpRecordSize(Instance instance) {
+        Klass reflectedKlass = java_lang_Class.asKlass(instance);
+        // dump instance record only for primitive type Class objects.
+        // all other Class objects are covered by writeClassDumpRecords.
+        if (reflectedKlass == null) {
+            return calculateInstanceDumpRecordSize(instance);
+        }
+        return 0;
+    }
+
+    private int calculateObjectArrayDumpRecordSize(ObjArray array) throws IOException {
+        int headerSize = getArrayHeaderSize(true);
+        final int length = calculateArrayMaxLength(array.getLength(),
+                headerSize,
+                OBJ_ID_SIZE,
+                "Object");
+        return headerSize + length * OBJ_ID_SIZE;
+    }
+
+    private int calculatePrimitiveArrayDumpRecordSize(TypeArray array) throws IOException {
+        int headerSize = getArrayHeaderSize(false);
+        TypeArrayKlass tak = (TypeArrayKlass) array.getKlass();
+        final int type = (int) tak.getElementType();
+        final String typeName = tak.getElementTypeName();
+        final long typeSize = getSizeForType(type);
+        final int length = calculateArrayMaxLength(array.getLength(),
+                                                   headerSize,
+                                                   typeSize,
+                                                   typeName);
+        return headerSize + (int)typeSize * length;
     }
 
     @Override
-    protected void writeHeapRecordEpilogue() throws IOException {
+    protected void writeHeapRecordPrologue(int size) throws IOException {
+        if (size == 0 || (currentSegmentStart > 0)) {
```

Comment:
Formatting nit: There is no need in internal brackets `()` around `currentSegmentStart > 0`.

Reply:
Hi Serguei, 
The currentSegmentStart here could be non-zero when  useSegmentHeapDump is false. so we need to check it here. 
Thanks!